src/treecstr.ml

```diff
@@ -0,0 +1,341 @@
+    if is_empty c then None else
+    Option.bind (fun c -> Some(c, nl_prio))
+      (try Some(IntPairSet.choose c.available)
+       with Not_found -> None)
+
+  let is_instantiated pair { available ; _ } = IntPairSet.mem pair available
+
+  let remove pair pool = { pool with
+                           available = IntPairSet.remove pair pool.available }
+
+  let export pair = pair
+
+  let instantiate vslot esl pool =
+    try
+      let ovs = IntMap.find esl pool.partial in
+      let available = IntPairSet.add (normalize (vslot, ovs))
+          pool.available in
+      { pool with available }
+    with Not_found -> { pool with
+                        partial = IntMap.add esl vslot pool.partial }
+
+end
+
+module FvConstraints : FvConstraintSig =
+struct
+  type t = (tvar array) IntMap.t
+
+  type cstr = int * tvar array
+
+  type decision = cstr cdecision
+
+  type out = int * tvar array
+
+  type data = tvar array
+
+  let pp_cstr oc (sl, vars) =
+    let module F = Format in
+    Format.fprintf oc "%d: {@[<h>%a@]}" sl
+      (Format.pp_print_list
+         ~pp_sep:(fun oc () -> Format.pp_print_string oc "; ") Print.pp_tvar)
+      (Array.to_list vars)
+
+  let pp oc available =
+    let module F = Format in
+    let pp_tvs : tvar list pp = F.pp_print_list
+        ~pp_sep:(fun oc () -> F.fprintf oc "; ")
+        Print.pp_tvar in
+    let ppit : (int * tvar array) pp = fun oc (a, b) ->
+      F.fprintf oc "@[(%d, %a)@]" a pp_tvs (Array.to_list b) in
+    F.fprintf oc "Fv constraints:@," ;
+    F.fprintf oc "@[<v>" ;
+    F.fprintf oc "@[available: %a@]@,"
+      (F.pp_print_list ppit)
+      (IntMap.bindings available) ;
+    F.fprintf oc "@]@."
+
+  let empty = IntMap.empty
+
+  let is_empty = IntMap.is_empty
+
+  let is_instantiated (sl, x) p =
+    try
+      let x' = IntMap.find sl p in
+      Array.equal Bindlib.eq_vars x x'
+    with Not_found -> false
+
+  let remove (sl, x) p =
+    try
+      let x' = IntMap.find sl p in
+      if Array.equal Bindlib.eq_vars x x' then
+        IntMap.remove sl p else p
+    with Not_found -> p
+
+  let instantiate slot vars pool =
+    IntMap.add slot vars pool
+
+  let export x = x
+
+  let score c =
+    try Some(IntMap.choose c, 1.)
+    with Not_found -> None
+end
+
+(** {3 Comparing constraints }*)
+
+(** Type making a module comparable. *)
+module type Scorable = sig
+  (** Type of the element to score. *)
+  type t
+
+  (** Result of a comparison. *)
+  type decision
+
+  (** [choose x] returns the action with the highest score. *)
+  val choose : t list -> decision
+end
+
+(** [MakeScorable(P)] returns a module containing a function to compare list
+    of elements of [P].  Acts as an {e extension} of [P]. *)
+module MakeScorable(BCP:BinConstraintPoolSig)
+  : (Scorable with type t := BCP.t and type decision := BCP.decision) =
+struct
+
+  open BCP
+
+  let score_gt s1 s2 = match (s1, s2) with
+    | None      , _          -> -1
+    | Some(_, x), None       -> max (int_of_float x) 1
```

Comment:
Can't we use a max function on floats?

More generally, is it useful to use floats?

Reply:
I use floats in one of the heuristics.

Reply:
score_gt is in fact a comparison function, right? It is not reflexive: None is strictly smaller than itself. Shouldn't you have None, None -> 0? It does look symmetric either.